return_type not correctly set for truediv operator in algebra.py

In certain case -- calculation between original array x and a result comes from truediv among x -- the calculation may cause NotImplementedError.
Also, other operators in _reflexive_operators are tested, only truediv has this error.

## Test code snippet is as follows:
```
def var(
    x,
    /,
    *,
    axis: int | tuple[int, ...] | None = None,
    correction: float = 0.0,
    keepdims: bool = False,
):
    s = elementwise(operator.truediv, x, x)
    v = x + s    # v = s + s works, only calculation between original array x and s from previous step causes error
    return v * v
```


## Complete callstack:
```
=================================== FAILURES ===================================
___________________________ test_lazyTensor_var[a0] ____________________________

a = array([[2, 4, 6, 8],
       [1, 3, 5, 7]])

    @pytest.mark.parametrize(
        "a",
        [
            (np.array([[2, 4, 6, 8], [1, 3, 5, 7]])),
        ],
    )
    def test_lazyTensor_var(a):
>       result = finch.var(a, axis=0)

tests/test_reduce_api_funcs.py:155: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/finch/interface/eager.py:206: in var
    return compute(lazy.var(x, axis=axis, correction=correction, keepdims=keepdims))
src/finch/interface/lazy.py:514: in var
    return v * v
src/finch/interface/lazy.py:61: in __mul__
    return multiply(self, defer(other))
src/finch/interface/lazy.py:384: in multiply
    return elementwise(operator.mul, defer(x1), defer(x2))
src/finch/interface/lazy.py:347: in elementwise
    new_element_type = return_type(f, *[x.element_type for x in args])
src/finch/algebra/algebra.py:164: in return_type
    return query_property(op, "__call__", "return_type", *args)
src/finch/algebra/algebra.py:78: in query_property
    return query_fn(obj, *args)
src/finch/algebra/algebra.py:206: in <lambda>
    else query_property(b, rmeth, "return_type", a),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <class 'NotImplementedType'>, attr = '__rmul__', prop = 'return_type'
args = (<class 'NotImplementedType'>,), query_fn = None
T = <class 'NotImplementedType'>, Ti = <class 'object'>

    def query_property(obj: type | Hashable, attr: str, prop: str, *args) -> Any:
        """Queries a property of an attribute of an object or class.  Properties can
        be overridden by calling register_property on the object or it's class.
    
        Args:
            obj: The object or class of object to query.
            attr: The attribute to query.
            prop: The property to query.
            args: Additional arguments to pass to the property.
    
        Returns:
            The value of the queried property.
    
        Raises:
            NotImplementedError: If the property is not implemented for the given type.
        """
        if not isinstance(obj, type):
            if isinstance(obj, Hashable):
                query_fn = _properties.get((obj, attr, prop))
                if query_fn is not None:
                    return query_fn(obj, *args)
            T = type(obj)
        else:
            T = obj
    
        for Ti in T.__mro__:
            query_fn = _properties.get((Ti, attr, prop))
            if query_fn is not None:
                return query_fn(obj, *args)
    
>       raise NotImplementedError(f"Property {prop} not implemented for {obj}")
E       NotImplementedError: Property return_type not implemented for <class 'NotImplementedType'>

src/finch/algebra/algebra.py:88: NotImplementedError
```

--- a/src/finch/algebra/algebra.py
+++ b/src/finch/algebra/algebra.py
@@ -153,7 +153,7 @@ def fill_value(arg: Any) -> Any:
         The fill value for the given argument.
 
     Raises:
-        NotImplementedError: If the fill value is not implemented for the given type.
+        AttributeError: If the fill value is not implemented for the given type.
     """
     if hasattr(arg, "fill_value"):
         return arg.fill_value
@@ -177,7 +177,7 @@ def element_type(arg: Any) -> type:
         The element type of the given tensor.
 
     Raises:
-        NotImplementedError: If the element type is not implemented for the given type.
+        AttributeError: If the element type is not implemented for the given type.
     """
     if hasattr(arg, "element_type"):
         return arg.element_type
@@ -203,7 +203,7 @@ def length_type(arg: Any) -> type:
         The length type of the given object.
 
     Raises:
-        NotImplementedError: If the length type is not implemented for the given type.
+        AttributeError: If the length type is not implemented for the given type.
     """
     if hasattr(arg, "length_type"):
         return arg.length_type
@@ -221,7 +221,7 @@ def shape_type(arg: Any) -> type:
         The shape type of the given object.
 
     Raises:
-        NotImplementedError: If the shape type is not implemented for the given type.
+        AttributeError: If the shape type is not implemented for the given type.
     """
     if hasattr(arg, "shape_type"):
         return arg.shape_type
@@ -254,9 +254,13 @@ def promote_type(a: Any, b: Any) -> type:
         The common type of the given arguments.
     """
     if hasattr(a, "promote_type"):
-        return a.promote_type(b)
+        res = a.promote_type(b)
+        if res is not NotImplemented:
+            return res
     if hasattr(b, "promote_type"):
-        return b.promote_type(a)
+        res = b.promote_type(a)
+        if res is not NotImplemented:
+            return res
     try:
         return query_property(a, "promote_type", "__attr__", b)
     except AttributeError:
@@ -311,7 +315,7 @@ _reflexive_operators = {
 }
 
 
-def _return_type_reflexive(meth):
+def _return_type_reflexive_method(meth):
     def _return_type_closure(a, b):
         if issubclass(b, StableNumber):
             return type(getattr(a(True), meth)(b(True)))
@@ -320,7 +324,22 @@ def _return_type_reflexive(meth):
     return _return_type_closure
 
 
+def _return_type_reflexive_operator(meth, rmeth):
+    def _return_type_closure(op, a, b):
+        if hasattr(a, meth):
+            try:
+                res = query_property(a, meth, "return_type", b)
+                if res is not type(NotImplemented):
+                    return res
+            except AttributeError:
+                pass
+        return query_property(b, rmeth, "return_type", a)
+
+    return _return_type_closure
+
+
 op: Callable
+
 
 for op, (meth, rmeth) in _reflexive_operators.items():
     (
@@ -328,17 +347,13 @@ for op, (meth, rmeth) in _reflexive_operators.items():
             op,
             "__call__",
             "return_type",
-            lambda op, a, b, meth=meth, rmeth=rmeth: query_property(
-                a, meth, "return_type", b
-            )
-            if hasattr(a, meth)
-            else query_property(b, rmeth, "return_type", a),
+            _return_type_reflexive_operator(meth, rmeth),
         ),
     )
 
     for t in StableNumber.__args__:
-        register_property(t, meth, "return_type", _return_type_reflexive(meth))
-        register_property(t, rmeth, "return_type", _return_type_reflexive(rmeth))
+        register_property(t, meth, "return_type", _return_type_reflexive_method(meth))
+        register_property(t, rmeth, "return_type", _return_type_reflexive_method(rmeth))
 
 
 _unary_operators: dict[Callable, str] = {
@@ -536,7 +551,7 @@ def type_min(t: type[T]) -> T:
         The minimum value of the given type.
 
     Raises:
-        NotImplementedError: If the minimum value is not implemented for the given type.
+        AttributeError: If the minimum value is not implemented for the given type.
     """
     if hasattr(t, "type_min"):
         return t.type_min()  # type: ignore[attr-defined]
@@ -554,7 +569,7 @@ def type_max(t: type[T]) -> T:
         The maximum value of the given type.
 
     Raises:
-        NotImplementedError: If the maximum value is not implemented for the given type.
+        AttributeError: If the maximum value is not implemented for the given type.
     """
     if hasattr(t, "type_max"):
         return t.type_max()  # type: ignore[attr-defined]
@@ -584,7 +599,7 @@ def init_value(op, arg) -> Any:
         The initial value for the given operation and type.
 
     Raises:
-        NotImplementedError: If the initial value is not implemented for the given type
+        AttributeError: If the initial value is not implemented for the given type
         and operation.
     """
     return query_property(op, "__call__", "init_value", arg)

--- a/tests/test_issues.py
+++ b/tests/test_issues.py
@@ -11,3 +11,15 @@ def test_issue_64():
     result = finch.compute(c).shape
     expected = (4, 2, 3)
     assert result == expected, f"Expected shape {expected}, got {result}"
+
+
+def test_issue_50():
+    x = finch.defer(np.array([[2, 4, 6, 8], [1, 3, 5, 7]]))
+    m = finch.defer(np.array([[1, 1, 1, 1], [1, 1, 1, 1]]))
+    n = finch.defer(
+        np.array([[2.0, 2.0, 2.0, 2.0], [2.0, 2.0, 2.0, 2.0]])
+    )  # Int -> Float caused return_type error
+    # If replaced above with below line, no error
+    # n = finch.defer(np.array([[2, 2, 2, 2], [2, 2, 2, 2]]))
+    o = finch.defer(np.array([[3, 3, 3, 3], [3, 3, 3, 3]]))
+    return finch.add(finch.add(finch.subtract(x, m), n), o)

